File: Disturbance.m
classdef wheelMobileRobot
    properties
        wheel_radius
        length
        prop_gain
        amplitude
        omega
    end
    
    methods
        function obj = wheelMobileRobot(wheel_radius, length, prop_gain)
            obj.wheel_radius = wheel_radius;
            obj.length = length;
            obj.prop_gain = prop_gain;
            obj.amplitude = 5;
            obj.omega = 0.2;
        end
        
        function [error_x, error_y, error_theta] = position_error(~, xd, yd, theta_d, x, y, theta)
            error_x = xd - x;
            error_y = yd - y;
            theta_d = atan2(error_y, error_x);
            theta_d = atan2(sin(theta_d), cos(theta_d));
            error_theta = theta_d - theta;
            error_theta = atan2(sin(error_theta), cos(error_theta));
        end
        
        function [theta_dot, velocity] = correction_error(obj, error_x, error_y, error_theta)
            mse = sqrt(error_x^2 + error_y^2);
            theta_dot = obj.prop_gain * error_theta;
            velocity = obj.prop_gain * mse;
        end
        
        function [phidotl, phidotr] = wheel_speeds(obj, velocity, theta_dot)
            velocity_vector = [velocity; theta_dot];
            wheel_radius_matrix = [obj.wheel_radius/2, obj.wheel_radius/2; ...
                                   -obj.wheel_radius/obj.length, obj.wheel_radius/obj.length];
            inverse_wheel_radius = inv(wheel_radius_matrix);
            wheel_speeds = inverse_wheel_radius * velocity_vector;
            phidotl = wheel_speeds(1);
            phidotr = wheel_speeds(2);
        end
        
        function [x_dot, y_dot, theta_dot] = diffmodel(obj, phidotl, phidotr, theta)
            x_dot = (obj.wheel_radius/2) * (phidotl + phidotr) * cos(theta);
            y_dot = (obj.wheel_radius/2) * (phidotl + phidotr) * sin(theta);
            theta_dot = (obj.wheel_radius/obj.length) * (phidotr - phidotl);
        end
        
        function [x, y, theta] = position_actuel(~, x, y, theta, x_dot, y_dot, theta_dot, dt)
            x = x + x_dot * dt;
            y = y + y_dot * dt;
            theta = theta + theta_dot * dt;
            theta = atan2(sin(theta), cos(theta)); % Normalization
        end
    end
end

% Main script
clear;
clc;

% Create robot object
robot = wheelMobileRobot(0.5, 2, 1.25);

% Variables
x = 0; y = 0; theta = 0;
amplitude = 5;
half_amplitude = amplitude / 2;
omega = 0.2;
positions = [x, y];
erreur_distance = [];
erreur_orientation = [];
dt = 0.1;
t_final = 60;
time_steps = 0:dt:t_final;

disp('Input the name of the trajectory you want the robot to describe: infinity, circular, target');
disp('YOU MUST INPUT EITHER: infinity, circular, or target');
input_trajectory = input('Chosen trajectory: ', 's');

switch input_trajectory
    case 'infinity'
        x_theorique = amplitude * sin(omega * time_steps);
        y_theorique = half_amplitude * sin(2 * omega * time_steps);
        
        for i = 1:length(time_steps)
            t = time_steps(i);
            x_target = amplitude * sin(t * omega);
            y_target = half_amplitude * sin(2 * t * omega);
            
            [err_x, err_y, err_theta] = robot.position_error(x_target, y_target, theta, x, y, theta);
            [theta_dot, vb] = robot.correction_error(err_x, err_y, err_theta);
            [phidotl, phidotr] = robot.wheel_speeds(vb, theta_dot);
            [x_dot, y_dot, theta_dot] = robot.diffmodel(phidotl, phidotr, theta);
            [x, y, theta] = robot.position_actuel(x, y, theta, x_dot, y_dot, theta_dot, dt);
            
            if mod(i, 10) == 0
                x = x + rand() * 0.2 - 0.1; % Perturbation in x
                y = y + rand() * 0.2 - 0.1; % Perturbation in y
                theta = theta + rand() * 0.2 - 0.1; % Perturbation in theta
            end
            
            positions = [positions; x, y];
            error_distance = sqrt((x - x_target)^2 + (y - y_target)^2);
            erreur_distance = [erreur_distance, error_distance];
            error_orientation = atan2(y_target - y, x_target - x) - theta;
            error_orientation = atan2(sin(error_orientation), cos(error_orientation));
            erreur_orientation = [erreur_orientation, error_orientation];
        end
        
    case 'circular'
        x_theorique = amplitude * sin(omega * time_steps);
        y_theorique = amplitude * cos(omega * time_steps);
        
        for i = 1:length(time_steps)
            t = time_steps(i);
            x_target = amplitude * sin(t * omega);
            y_target = amplitude * cos(t * omega);
            
            [err_x, err_y, err_theta] = robot.position_error(x_target, y_target, theta, x, y, theta);
            [theta_dot, vb] = robot.correction_error(err_x, err_y, err_theta);
            [phidotl, phidotr] = robot.wheel_speeds(vb, theta_dot);
            [x_dot, y_dot, theta_dot] = robot.diffmodel(phidotl, phidotr, theta);
            [x, y, theta] = robot.position_actuel(x, y, theta, x_dot, y_dot, theta_dot, dt);
            
            if mod(i, 10) == 0
                x = x + rand() * 0.2 - 0.1; % Perturbation in x
                y = y + rand() * 0.2 - 0.1; % Perturbation in y
                theta = theta + rand() * 0.2 - 0.1; % Perturbation in theta
            end
            
            positions = [positions; x, y];
            error_distance = sqrt((x - x_target)^2 + (y - y_target)^2);
            erreur_distance = [erreur_distance, error_distance];
            error_orientation = atan2(y_target - y, x_target - x) - theta;
            error_orientation = atan2(sin(error_orientation), cos(error_orientation));
            erreur_orientation = [erreur_orientation, error_orientation];
        end
    otherwise
        disp('The given trajectory name is not known');
end

% Plotting results
positions = positions(2:end, :);
x_vals = positions(:, 1);
y_vals = positions(:, 2);

figure;
subplot(1, 2, 1);
plot(x_vals, y_vals, 'b', 'DisplayName', 'Robot Trajectory');
hold on;
plot(x_theorique, y_theorique, 'm--', 'DisplayName', 'Theoretical Trajectory');
xlabel('x (m)');
ylabel('y (m)');
legend;
title('Trajectory Tracking');
grid on;
axis equal;

subplot(1, 2, 2);
plot(time_steps(1:end-1), erreur_distance, 'b', 'DisplayName', 'Distance Error');
hold on;
plot(time_steps(1:end-1), erreur_orientation, 'r', 'DisplayName', 'Orientation Error');
xlabel('Time (s)');
ylabel('Error');
legend;
title('Tracking Errors');
grid on;

File: disturbance.py
import matplotlib.pyplot as plt
import numpy as np


class wheelMobileRobot():
    def __init__(self, wheel_radius, length, prop_gain):
        self.wheel_radius = wheel_radius
        self.length = length
        self.prop_gain = prop_gain
        self.amplitude = 5
        self.omega = 0.2



    """fonction used to compute erro """
    def position_error(self,xd, yd, theta_d, x, y, theta):
        error_x = xd - x 
        error_y = yd - y 
        theta_d = np.arctan2(error_y,error_x)
        theta_d = np.arctan2(np.sin(theta_d), np.cos(theta_d))
        error_theta = theta_d - theta
        error_theta = np.arctan2(np.sin(error_theta), np.cos(error_theta))

        return error_x, error_y, error_theta
    

    """compute vb and theta_dot then apply the gain """
    def correction_error(self, error_x, error_y, error_theta):
        mse = np.sqrt(error_x**2 + error_y**2)
        theta_dot = self.prop_gain*(error_theta)
        velocity = self.prop_gain * mse

        return theta_dot, velocity
    

    """This function returns the left and right wheel speeds 
    phidotl and phidotr """
    def wheel_speeds(self, velocity, theta_dot):
        velocity_vector = np.array([[velocity],[theta_dot]])
        wheel_radius_matrix = np.array([[self.wheel_radius/2, self.wheel_radius/2], [-self.wheel_radius/self.length, self.wheel_radius/self.length]])
        inverse_wheel_radius = np.linalg.inv(wheel_radius_matrix)
        wheel_speeds = np.dot(inverse_wheel_radius, velocity_vector)
        phidotl = wheel_speeds[0,0]
        phidotr = wheel_speeds[1,0]

        return phidotl, phidotr
    

    """Compute theta, the velocity on x (x dx/dt) and y (y dy/dt) direction with respect to the time"""
    def diffmodel(self, phidotl, phidotr, theta):
        #interpolate the points in the time array with
        #those in the speed array

        x_dot = (self.wheel_radius/2) * (phidotl  + phidotr ) * np.cos(theta)

        y_dot = (self.wheel_radius/2) *  (phidotl  + phidotr) *  np.sin(theta)

        theta_dot = (self.wheel_radius/self.length)*(phidotr - phidotl)

        v_world = [x_dot, y_dot, theta_dot]

        #v_world = np.array(v_world)

        return x_dot, y_dot, theta_dot
    

    def position_actuel(self, x, y, theta, x_dot, y_dot, theta_dot, dt):
        x += x_dot * dt 
    
        y += y_dot * dt 
    
        theta += theta_dot * dt
    
        theta = np.arctan2(np.sin(theta), np.cos(theta)) #normalisation
    
        return x, y, theta



def main():
    #class object
    robot = wheelMobileRobot(wheel_radius= 0.5 , length= 2, prop_gain=1.25)
    #variables
    x, y, theta = 0.0, 0.0, 0.0
    amplitude = 5
    half_amplitude = amplitude/2
    omega = 0.2
    positions = [(x,y)]
    error_distance_list = []
    error_orientation_list = []
    theta_list = []
    dt = 0.1
    t_final = 60
    time_steps = np.arange(0, t_final, dt)

    print("Input the name of the trajectory you want the robot to describe: infinity, circular, target")
    print("YOU MUST INPUT EITHER: infinity, circular or target")
    input_trajectory = input("Chosen trajectory:")

    if input_trajectory == "infinity":
        
        x_theoretical = amplitude * np.sin(omega*time_steps)
        y_theoretical = half_amplitude * np .sin(2*omega*time_steps)
        
        
        for i,t in enumerate(time_steps):
            x_target = amplitude * np.sin(t*omega)
            y_target = half_amplitude * np.sin(2*t*omega)

             #position error calculation 
            err_x, err_y, err_theta = robot.position_error(x_target, y_target, theta, x, y, theta)

            #velocity and theta_dot 
            theta_dot, vb = robot.correction_error(err_x, err_y, err_theta)
            phidotl, phidotr = robot.wheel_speeds(vb, theta_dot)

            #using the motion equations
            x_dot, y_dot, theta_dot = robot.diffmodel(phidotl,phidotr,theta)

            #actual position and orientation
            x, y, theta = robot.position_actuel(x,y,theta, x_dot, y_dot, theta_dot, dt)
            
            #Applying a disturbance to the system
              # Introduction of perturbations every 10 iterations
            if i % 10 == 0:
                x += np.random.uniform(-0.1, 0.1)  # Perturbation sur x
                y += np.random.uniform(-0.1, 0.1)  # Perturbation sur y
                theta += np.random.uniform(-0.1, 0.1)  # Perturbation sur theta

            #saving present position
            positions.append((x, y))

            error_distance = np.sqrt((x - x_target)**2 + (y - y_target)**2)
            error_distance_list.append(error_distance)

            error_orientation = np.arctan2(y_target - y, x_target - x) - theta
            error_orientation = np.arctan2(np.sin(error_orientation), np.cos(error_orientation))
            error_orientation_list.append(error_orientation)

    elif input_trajectory == "target":
        
        x_target = int(input("Please input the coordinates on the x-axis:"))
        y_target = int(input("Please input the coordinates on the y-axis:"))
        theta_target = int(input("please input the coodinates of theta:"))
        
        x_theoretical = np.array([x_target])
        y_theoretical = np.array([y_target])

        for i,t in enumerate(time_steps):
            
            #position error calculation 
            err_x, err_y, err_theta = robot.position_error(x_target, y_target, theta, x, y, theta)

            #velocity and theta_dot 
            theta_dot, vb = robot.correction_error(err_x, err_y, err_theta)
            phidotl, phidotr = robot.wheel_speeds(vb, theta_dot)

            #using the motion equations
            x_dot, y_dot, theta_dot = robot.diffmodel(phidotl,phidotr,theta)

            #actual position and orientation
            x, y, theta = robot.position_actuel(x,y,theta, x_dot, y_dot, theta_dot, dt)
            
            if i % 10 == 0:
                x += np.random.uniform(-0.1, 0.1)  # Perturbation sur x
                y += np.random.uniform(-0.1, 0.1)  # Perturbation sur y
                theta += np.random.uniform(-0.1, 0.1)  # Perturbation sur theta

            #saving present position
            positions.append((x, y))

            error_distance = np.sqrt((x - x_target)**2 + (y - y_target)**2)
            error_distance_list.append(error_distance)

            error_orientation = np.arctan2(y_target - y, x_target - x) - theta
            error_orientation = np.arctan2(np.sin(error_orientation), np.cos(error_orientation))
            error_orientation_list.append(error_orientation)

        
    elif input_trajectory == "circular":
        x_theoretical = amplitude * np.sin(omega * time_steps)
        y_theoretical = amplitude * np.cos(omega * time_steps)
        
        for i,t in enumerate(time_steps):
            x_target = amplitude *  np.sin(t*omega)
            y_target = amplitude * np.cos(t*omega)
            #position error calculation 
            err_x, err_y, err_theta = robot.position_error(x_target, y_target, theta, x, y, theta)

            #velocity and theta_dot 
            theta_dot, vb = robot.correction_error(err_x, err_y, err_theta)
            phidotl, phidotr = robot.wheel_speeds(vb, theta_dot)

            #using the motion equations
            x_dot, y_dot, theta_dot = robot.diffmodel(phidotl,phidotr,theta)

            #actual position and orientation
            x, y, theta = robot.position_actuel(x,y,theta, x_dot, y_dot, theta_dot, dt)
            
            if i % 10 == 0:
                x += np.random.uniform(-0.1, 0.1)  # Perturbation sur x
                y += np.random.uniform(-0.1, 0.1)  # Perturbation sur y
                theta += np.random.uniform(-0.1, 0.1)  # Perturbation sur theta

            #saving present position
            positions.append((x, y))

            error_distance = np.sqrt((x - x_target)**2 + (y - y_target)**2)
            error_distance_list.append(error_distance)

            error_orientation = np.arctan2(y_target - y, x_target - x) - theta
            error_orientation = np.arctan2(np.sin(error_orientation), np.cos(error_orientation))
            error_orientation_list.append(error_orientation)
    
    else:
        print("The given trajectory name is not known")
    positions = np.array(positions)
    x_vals, y_vals = positions[:, 0], positions[:, 1]
        # Display of the trajectory of the real robot an the theorical trajectory
    print(x_vals.shape[0])
    if x_vals.shape[0] >= 2:
        plt.figure(figsize=(12, 6))

        # graph of the trajectory
        plt.subplot(1, 2, 1)
        plt.plot(x_vals, y_vals, label="real robot trajectory", color="blue")
        plt.plot(x_theoretical, y_theoretical, label="theorical trajectory", color="pink", linestyle=":")
        plt.scatter(x_vals[0], y_vals[0], color='red', label='initial position')
        plt.scatter(x_theoretical[-1], y_theoretical[-1], color='blue', label='target Position')
        plt.xlabel("Position x (m)")
        plt.ylabel("Position y (m)")
        plt.legend()
        plt.title("trajectory of the differential drive robot with Sliding Mode Control")
        plt.grid()
        plt.axis("equal")


        # Graph of errors
        plt.subplot(1, 2, 2)
        plt.plot(time_steps, error_distance_list, label="distance error", color="blue")
        plt.plot(time_steps, error_orientation_list, label="orientation error", color="orange")
        plt.xlabel("Time (s)")
        plt.ylabel("Error")
        plt.legend()
        plt.title("Evolution of errors with respect to the time")
        plt.grid()

        plt.tight_layout()
        plt.show()



if __name__ == "__main__":
    main()
